### PR TITLE
gbConf: Use shared memory instead of from the data segment

### PR DESCRIPTION
**Pull request by lxbsz**
### What does this PR achieve? Why do we need it?

Since we have make the cli and server routines into 2 different
processes in gluster-blockd daemon, the gbConf data will also
be have different copies. If the gbConf data is changed by the
dyn-config thread in the cli(parent) process, the new changes won't
be visible for the server(child) process.

Resolution:
----------
For the gbConf data, use the shared memory instead of the in the
data segment.

### CODE REVIEW

**Comment by lxbsz** [comment]
@pkalever @amarts 
Rebased to the latest, please review.
Thanks.

File: cli/gluster-block.c [hunk 1]
```diff
@@ -1029,13 +1029,18 @@ main(int argc, char *argv[])
...
   int json = GB_JSON_NONE;
+  int ret;
 
 
   if (argc <= 1) {
     glusterBlockHelp();
     goto fail;
   }
 
+  if (initGbConfig()) {
```
Comment (pkalever) [inline cli/gluster-block.c]: I don't see a reason why we need to make the gbConf go in shared memory for cli (gluster-block process)
Did you just called it coz you might want to make the lock initialization and gbConf allocation common ?
I don't see any critical issue by doing this, but just raising the hand.

File: daemon/gluster-blockd.c [hunk 2]
```diff
@@ -19,6 +19,7 @@
 # include  <sys/utsname.h>
 # include  <linux/version.h>
 # include  <sys/wait.h>
+# include  <sys/mman.h>
```
Comment (pkalever) [inline daemon/gluster-blockd.c]: may be not need here

File: cli/gluster-block.c (hunk 1 above)
Comment (lxbsz) [reply cli/gluster-block.c]: @pkalever 
Me too, it make no sense here :-), and will fix it.
Thanks.

File: daemon/gluster-blockd.c (hunk 2 above)
Comment (lxbsz) [reply daemon/gluster-blockd.c]: Will remove it.
Thanks.

**Comment by lxbsz** [comment]
> Tested this by adding some custom logs in daemons child and parent paths, and they seem to work well with this patch.
> 
@pkalever 

No need to add extra logs to test this.

You can just try to change the logLevel in the /etc/sysconfig/gluster-blockd, INFO --> DEBUG, and then DEBUG --> INFO. And at the same time just test it by ./tests/basic.t 

When it is DEBUG you can see many DEBUG logs like the targetcli outputs and many other from the rpc/*.c files.

When it is INFO, all that logs will be disapear.

 Thanks.

> @lxbsz just added some comments, please take a look. Over all looks good to me.

**Comment by pkalever** [comment]
> > Tested this by adding some custom logs in daemons child and parent paths, and they seem to work well with this patch.
> 
> @pkalever
> 
> No need to add extra logs to test this.

@lxbsz My tests intention was to see if the server process is updating the gbConf and will the same is reflected to child process. i.e, is server and child looking at the same gbConf.

```
[root@localhost gluster-block]# git di 
diff --git a/rpc/block_svc_routines.c b/rpc/block_svc_routines.c
index 0422319..3156c6d 100644
--- a/rpc/block_svc_routines.c
+++ b/rpc/block_svc_routines.c
@@ -3869,6 +3869,9 @@ block_create_cli_1_svc_st(blockCreateCli *blk, struct svc_req *rqstp)
   bool *resultCaps = NULL;
 
 
+  LOG("mgmt", GB_LOG_CRIT, "pid: %u, LOG-LEVEL: %s", getpid(), LogLevelLookup[gbConf->logLevel]);
+
+
   LOG("mgmt", GB_LOG_INFO,
       "create cli request, volume=%s blockname=%s mpath=%d blockhosts=%s "
       "authmode=%d size=%lu, rbsize=%d", blk->volume, blk->block_name, blk->mpath,
@@ -4288,6 +4291,7 @@ block_create_common(blockCreate *blk, char *rbsize, char *volServer, char *prio_
   size_t i;
   bool prioCap = false;
 
+  LOG("mgmt", GB_LOG_CRIT, "pid: %u, LOG-LEVEL: %s", getpid(), LogLevelLookup[gbConf->logLevel]);
 
   LOG("mgmt", GB_LOG_INFO,
       "create request, volume=%s volserver=%s blockname=%s blockhosts=%s "
[root@localhost gluster-block]# 
```

This was one tests,
-> Check the LogLevel on changing the configfile
- where if I test without your patch, parent will get new LogLevel (TRACE) and Child will still see INFO (old)
- With your patch, parent and child see TRACE

Hope that makesense for you now, why I had to add extra log.

Thanks!

**Comment by lxbsz** [comment]
> > > Tested this by adding some custom logs in daemons child and parent paths, and they seem to work well with this patch.
> > 
> > 
> > @pkalever
> > No need to add extra logs to test this.
> 
> @lxbsz My tests intention was to see if the server process is updating the gbConf and will the same is reflected to child process. i.e, is server and child looking at the same gbConf.
> 
> ```
> [root@localhost gluster-block]# git di 
> diff --git a/rpc/block_svc_routines.c b/rpc/block_svc_routines.c
> index 0422319..3156c6d 100644
> --- a/rpc/block_svc_routines.c
> +++ b/rpc/block_svc_routines.c
> @@ -3869,6 +3869,9 @@ block_create_cli_1_svc_st(blockCreateCli *blk, struct svc_req *rqstp)
>    bool *resultCaps = NULL;
>  
>  
> +  LOG("mgmt", GB_LOG_CRIT, "pid: %u, LOG-LEVEL: %s", getpid(), LogLevelLookup[gbConf->logLevel]);
> +
> +
>    LOG("mgmt", GB_LOG_INFO,
>        "create cli request, volume=%s blockname=%s mpath=%d blockhosts=%s "
>        "authmode=%d size=%lu, rbsize=%d", blk->volume, blk->block_name, blk->mpath,
> @@ -4288,6 +4291,7 @@ block_create_common(blockCreate *blk, char *rbsize, char *volServer, char *prio_
>    size_t i;
>    bool prioCap = false;
>  
> +  LOG("mgmt", GB_LOG_CRIT, "pid: %u, LOG-LEVEL: %s", getpid(), LogLevelLookup[gbConf->logLevel]);
>  
>    LOG("mgmt", GB_LOG_INFO,
>        "create request, volume=%s volserver=%s blockname=%s blockhosts=%s "
> [root@localhost gluster-block]# 
> ```
> This was one tests,
> -> Check the LogLevel on changing the configfile
> 
> * where if I test without your patch, parent will get new LogLevel (TRACE) and Child will still see INFO (old)
> * With your patch, parent and child see TRACE
> 
> Hope that makesense for you now, why I had to add extra log.

Yeah, this make sense and is cool.

Thanks.

**Comment by lxbsz** [comment]
@pkalever 
Update it, please review.
Thanks.

File: utils/utils.c [hunk 3]
```diff
@@ -37,6 +30,66 @@ const char *argp_program_version = ""                                 \
...
+      return 0;
+
+  if (gbCtx == GB_CLI_MODE) {
+    if (GB_ALLOC_N(gbConf, sizeof(gbConf)) < 0) {
+      ret = errno;
+      MSG(stderr, "Failed to allocate memory for gbConf %s", strerror (errno));
+      return ret;
+    }
+
+    pthread_mutex_init(&gbConf->lock, NULL);
```
Comment (pkalever) [inline utils/utils.c]: @lxbsz where are we initializing the other members of gbConf  for GB_CLI_MODE ?

You have removed:

```
struct gbConf gbConf = {
	  .lock = PTHREAD_MUTEX_INITIALIZER,
	  .glfsLruCount = LRU_COUNT_DEF,
	  .logLevel = GB_LOG_INFO,
	  .logDir = GB_LOGDIR_DEF,
	  .cliTimeout = CLI_TIMEOUT_DEF
};

```

Comment (lxbsz) [reply utils/utils.c]: Ah, sorry, should remove the following "return 0"

**Comment by lxbsz** [comment]
@pkalever 
Update it.
Thanks.

**Comment by lxbsz** [comment]
> @lxbsz patch is not working as expected :-(
> 
> Failed with below test:
> 
> 1. change loglevel to TRACE
>    [2019-04-26 13:21:15.679441] CRIT: logLevel now is TRACE [at utils.c+108 :]
> 2. Now let the LOGLEVEL be TRACE and change GB_GLFS_LRU_COUNT=10
>    [2019-04-26 13:21:34.374815] CRIT: logLevel now is INFO [at utils.c+108 :]
>    [2019-04-26 13:21:34.374933] CRIT: glfsLruCount now is 10 [at lru.c+72 :]
> 
> Which is wrong!!
> 
> 1. Now let the  LOGLEVEL be TRACE and GB_GLFS_LRU_COUNT=10, change GB_CLI_TIMEOUT=100
>    [2019-04-26 13:24:25.966297] CRIT: logLevel now is INFO [at utils.c+108 :]
>    [2019-04-26 13:24:25.966393] CRIT: glfsLruCount now is 5 [at lru.c+72 :]
> 
> Which is wrong.
> 
> This means we are not persisting the previous values. And on inotify just caring for newly edited values.
> 

@pkalever 

BTW, did you change any other code ? I have tried 10 times locally, all were working as expected with your steps:

```
[2019-04-28 02:47:52.104104] CRIT: logLevel now is TRACE [at utils.c+108 :<glusterBlockSetLogLevel>]
[2019-04-28 02:47:52.104162] CRIT: glfsLruCount now is 5 [at lru.c+72 :<glusterBlockSetLruCount>]
[2019-04-28 02:47:52.112534] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:47:57.869107] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:00.664798] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:05.548109] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:06.548692] CRIT: logLevel now is TRACE [at utils.c+108 :<glusterBlockSetLogLevel>]
[2019-04-28 02:48:06.548742] CRIT: glfsLruCount now is 10 [at lru.c+72 :<glusterBlockSetLruCount>]
[2019-04-28 02:48:06.548764] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:06.548798] CRIT: logLevel now is TRACE [at utils.c+108 :<glusterBlockSetLogLevel>]
[2019-04-28 02:48:06.548812] CRIT: glfsLruCount now is 10 [at lru.c+72 :<glusterBlockSetLruCount>]
[2019-04-28 02:48:06.548825] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:09.562048] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:15.437533] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:19.931159] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]
[2019-04-28 02:48:19.931279] CRIT: logLevel now is TRACE [at utils.c+108 :<glusterBlockSetLogLevel>]
[2019-04-28 02:48:19.931311] CRIT: glfsLruCount now is 10 [at lru.c+72 :<glusterBlockSetLruCount>]
[2019-04-28 02:48:19.940056] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :<glusterBlockDynConfigStart>]

```
Did I missed something ?

Thanks.
BRs

**Comment by lxbsz** [comment]
@pkalever 
Test it in Fedora 29 again, and also the same with above.

Could you try again ?

Thanks.

File: utils/utils.c [hunk 4]
```diff
@@ -459,6 +476,15 @@ initLogDirAndFiles(char *newLogDir)
    */
   if (newLogDir) {
     logDir = newLogDir;
+    LOCK(gbConf->lock);
+    if (!strcmp(logDir, gbConf->logDir)) {
+      UNLOCK(gbConf->lock);
+      LOG(dom, GB_LOG_ERROR,
```
Comment (pkalever) [inline utils/utils.c]: @lxbsz 
should this be GB_LOG_DEBUG ?

Comment (lxbsz) [reply utils/utils.c]: Yeah, will fix it.
Thanks.

**Comment by lxbsz** [comment]
> @lxbsz I hit something wrong again:
> 
> I have first changed the loglevel to TRACE:
> 
> [2019-04-29 11:11:36.102162] CRIT: logLevel now is TRACE [at utils.c+126 :]
> 
> Then uncomented the LRUCOUNT
> 
> [2019-04-29 11:11:43.647928] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :]
> [2019-04-29 11:11:43.648127] CRIT: logLevel now is INFO [at utils.c+126 :]
> 
> And I don't see an log about LRUCOUNT.
> 

BTW, did you change the default value ? or just keep it as 5 ?

If you didn't change it, and at the same time the logLevel is INFO, there shouldn't any logs about this.

> My question is why the logLevel is set back to INFO when I didn't comment it or change the TRACE level to something else.

If you still could hit the issue, it should be dyn-config's bug.
As you know, after every time it reloaded from the config file and set to gbConf, all the options' value in options list will be set to DEF again, and when the next time to reload the config file, if the options are commented, it will set the DEF values to gbConf.

Thanks.
> 
> HOwever this is not 100% reproducable, please keep playing with the configfile on F29 and see if you hit it again.
> 
> Thanks!

**Comment by pkalever** [comment]
> > @lxbsz I hit something wrong again:
> > I have first changed the loglevel to TRACE:
> > [2019-04-29 11:11:36.102162] CRIT: logLevel now is TRACE [at utils.c+126 :]
> > Then uncomented the LRUCOUNT
> > [2019-04-29 11:11:43.647928] DEBUG: event->mask: 0x2 [at dyn-config.c+572 :]
> > [2019-04-29 11:11:43.648127] CRIT: logLevel now is INFO [at utils.c+126 :]
> > And I don't see an log about LRUCOUNT.
> 
> BTW, did you change the default value ? or just keep it as 5 ?

IMO, It doesn't matter here. I might have just uncommented and saved the file. 


> 
> If you didn't change it, and at the same time the logLevel is INFO, there shouldn't any logs about this.

Nope, The loglevel was untouched, it was TRACE as part of previous time change. Hence I don't expect it to change to INFO when the file has it set to TRACE.

> 
> > My question is why the logLevel is set back to INFO when I didn't comment it or change the TRACE level to something else.
> 
> If you still could hit the issue, it should be dyn-config's bug.
> As you know, after every time it reloaded from the config file and set to gbConf, all the options' value in options list will be set to DEF again, and when the next time to reload the config file, if the options are commented, it will set the DEF values to gbConf.

I didn't understand this part completely, are you saying this is expected ? or a bug which needs to be fixed ?

If you mean, we need something else to be fixed, can you please sendout a patch as part of this PR, as I don't feel this PR is complete without that missing piece.  

Thanks!




> 
> Thanks.
> 
> > HOwever this is not 100% reproducable, please keep playing with the configfile on F29 and see if you hit it again.
> > Thanks!

**Comment by lxbsz** [comment]
@pkalever 
Please test it again.

The root cause is that:

In-case if the editor (vim) follows write to a new file (.swp, .tmp ..) and move it to actual file name later. There is a window, where we will encounter one case that the file data is not flushed to the disk, so in
another process(here) when reading it will be empty.

Will try to read it again by waiting 1 second.
Thanks.

File: utils/dyn-config.c [hunk 5]
```diff
@@ -293,6 +296,23 @@ glusterBlockReadConfig(gbConfig *cfg, ssize_t *len)
     *len += m;
   }
 
+  /*
+   * In-case if the editor (vim) follows write to a new file (.swp, .tmp ..)
+   * and move it to actual file name later. There is a window, where we will
+   * encounter one case that the file data is not flushed to the disk, so in
+   * another process(here) when reading it will be empty.
+   *
+   * Let just wait and try again.
```
Comment (pkalever) [inline utils/dyn-config.c]: @lxbsz we have handled the emty file case, but in our chat you also mentioned that we see errno = 2, in some cases, can we also handle that part ?

Thanks!

**Comment by lxbsz** [comment]
The test logs:

```
[2019-04-29 14:20:03.708188] DEBUG: event->mask: 0x2 [at dyn-config.c+592 :<glusterBlockDynConfigStart>]
[2019-04-29 14:20:06.722769] DEBUG: event->mask: 0x2 [at dyn-config.c+592 :<glusterBlockDynConfigStart>]
[2019-04-29 14:20:08.438917] DEBUG: event->mask: 0x2 [at dyn-config.c+592 :<glusterBlockDynConfigStart>]
[2019-04-29 14:20:08.439038] DEBUG: The pagecache is not flushed yet, will try to read it again [at dyn-config.c+310 :<glusterBlockReadConfig>]
[2019-04-29 14:20:09.439288] DEBUG: No changes to current logLevel: DEBUG, skipping it. [at utils.c+120 :<glusterBlockSetLogLevel>]
[2019-04-29 14:20:09.439315] DEBUG: No changes to current glfsLruCount: 5, skipping it. [at lru.c+72 :<glusterBlockSetLruCount>]
[2019-04-29 14:20:09.439328] DEBUG: event->mask: 0x2 [at dyn-config.c+592 :<glusterBlockDynConfigStart>]
[2019-04-29 14:20:09.439337] DEBUG: event->mask: 0x2 [at dyn-config.c+592 :<glusterBlockDynConfigStart>]
```

File: utils/dyn-config.c (hunk 5 above)
Comment (lxbsz) [reply utils/dyn-config.c]: @pkalever 
Yeah, this have included this case already.
The errno == 2 case and empty case have the same root cause as above memtioned.

Thanks.

Comment (lxbsz) [reply utils/dyn-config.c]: This is why I added the "retry: " tag just before the fopen.

File: utils/dyn-config.c [hunk 6]
```diff
@@ -274,6 +276,7 @@ glusterBlockReadConfig(gbConfig *cfg, ssize_t *len)
   *len = 0;
   p = buf;
   while ((m = getline(&line, &n, fp)) != -1) {
+    empty = false;
```
Comment (pkalever) [inline utils/dyn-config.c]: @lxbsz 
Also, why we need an extra variable here ?
Can we directly check if buf is empty ?

Comment (lxbsz) [reply utils/dyn-config.c]: @pkalever 
Yeah, fixed it.
Thanks.

File: utils/dyn-config.c [hunk 7]
```diff
@@ -293,6 +294,22 @@ glusterBlockReadConfig(gbConfig *cfg, ssize_t *len)
...
+   * another process(here) when reading it will be empty.
+   *
+   * Let just wait and try again.
+   */
+  if (*len == 0) {
+    LOG("mgmt", GB_LOG_DEBUG,
+        "The pagecache is not flushed yet, will try to read it again");
+    fclose(fp);
+    sleep(1);
+    goto retry;
```
Comment (pkalever) [inline utils/dyn-config.c]: @lxbsz, should there be some limit to this retry ?
May be 5 times like we do in https://github.com/open-iscsi/tcmu-runner/pull/506/commits/f7b766b5e595ade619a023a4b1e1f0e1754c6d39 ?

Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, make sense.

**Comment by pkalever** [comment]
Thanks @lxbsz, Merged now :-)

**Pull request merged**